File: leetcode/medium/1353_maxEvents.ts
function maxEvents_0(events: number[][]): number {
    events.sort((a, b) => a[1] - b[1]);

    let count: number = 0;
    let now: number = 1;

    for (let i: number = 0; i < events.length; ++i) {
        const [s, e] = events[i];
        if (now > e) {
            continue;
        }
        count += 1;
        if (now >= s && now <= e) {
            now += 1;
            continue;
        }
        if (now < s) {
            now = s+1;
        }
    }

    return count;
};

function maxEvents(events: number[][]): number {
    const data: number[][] = events.map(([s, e], id) => [s, e, id]);
    const visited: number[] = new Array(data.length).fill(0);

    const byStart: number[][] = [...data].sort((a, b) => {
        if (a[0] === b[0]) return a[1] - b[1];
        return a[0] - b[0];
    });

    const byEnd: number[][] = [...data].sort((a, b) => {
        if (a[1] === b[1]) return a[0] - b[0];
        return a[1] - b[1];
    });

    let i: number = 0;
    let j: number= 0;

    let now: number = 0;
    let count: number = 0;

    while (i < byStart.length || j < byEnd.length) {
        while (i < byStart.length && (visited[byStart[i][2]] || byStart[i][1] < now)) {
            i += 1;
        }
        while (j < byEnd.length && (visited[byEnd[j][2]] || byEnd[j][1] < now)) {
            j += 1;
        }

        if (j < byEnd.length && now < byEnd[j][0] && i < byStart.length && now >= byStart[i][0]) {
            visited[byStart[i][2]] = 1;
            now += 1;
            i += 1;
            count += 1;
            continue;
        }

        if (j < byEnd.length && now >= byEnd[j][0]) {
            visited[byEnd[j][2]] = 1;
            now += 1;
            j += 1;
            count += 1;
            continue;
        }

        now += 1;

    }

    return count;
};

/*

1 2 3 4 5 6
1 1 -1 -1
1 1 -1 -1
1      -1

*/

const test = () => {
    const params = [
        {
            input: {
                events: [[1,5],[1,5],[1,5],[2,3],[2,3]],
            },
            output: 5,
        },
        {
            input: {
                events: [[1,2],[2,3],[3,4]],
            },
            output: 3,
        },
        {
            input: {
                events: [[1,2],[2,3],[3,4],[1,2]]
            },
            output: 4,
        },
    ];

    params.forEach(({input, output}) => {
        const { events } = input;
        const result = maxEvents(events);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

File: leetcode/medium/1353_maxEvents.py
import sys
sys.setrecursionlimit(1500)
from typing import List
import json
from collections import deque, defaultdict
import heapq

class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        n = len(events)
        events.sort()
        eventsQueue = []
        id = 0
        now = 1
        res = 0

        while id < n or eventsQueue:
            while id < n and events[id][1] < now:
                id += 1

            if not eventsQueue and id < n:
                now = max(now, events[id][0])

            while eventsQueue and eventsQueue[0] < now:
                heapq.heappop(eventsQueue)

            while id < n and events[id][0] == now:
                heapq.heappush(eventsQueue, events[id][1])
                id += 1

            if eventsQueue and eventsQueue[0] >= now:
                heapq.heappop(eventsQueue)
                res += 1

            now += 1


        return res



def test ():
    params = [
        {
            'input': [[1,5],[1,5],[1,5],[2,3],[2,3]],
            'output': 5,
        },
        {
            'input': [[1,2],[2,3],[3,4]],
            'output': 3,
        },
        {
            'input': [[1,2],[2,3],[3,4],[1,2]],
            'output': 4,
        },
    ]
    solution = Solution()

    for param in params:
        events = param['input']
        result = solution.maxEvents(events)
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()
